common/singleton.h

```diff
@@ -40,38 +40,34 @@ namespace Common {
  */
 template<class T>
 class Singleton : NonCopyable {
-private:
-	Singleton<T>(const Singleton<T> &);
-	Singleton<T> &operator=(const Singleton<T> &);
-
-	/**
-	 * The default object factory used by the template class Singleton.
-	 * By specialising this template function, one can make a singleton use a
-	 * custom object factory. For example, to support encapsulation, your
-	 * singleton class might be pure virtual (or "abstract" in Java terminology),
-	 * and you specialise makeInstance to return an instance of a subclass.
-	 */
-	//template<class T>
-#if defined(__WINS__)
-//FIXME verify if __WINS__ needs this still
 public:
-#endif
-	static T *makeInstance() {
-		return new T();
+	static T &instance() {
+		// TODO: We don't leak, but the destruction order is nevertheless
+		// semi-random. If we use multiple singletons, the destruction
+		// order might become an issue. There are various approaches
+		// to solve that problem, but for now this is sufficient
+		static T *_singleton = new T();
+		return *_singleton;
 	}
 
-	static void destroyInstance() {
-		delete _singleton;
-		_singleton = 0;
+	static void destroy() {
+		delete &instance();
```

Comment:
This means that the object gets freed while _singleton still points at it.
Next call to `instance()` after destroy will just break everything.

Reply:
How about removing `destroy` altogether, except for `OptionalSingleton`, and using a static object instead of a pointer, then it will be auto-released on termination?

Reply:
This is too risky. Is it a real use-case to call `instance()` after `destroy()`? Does it make sense at all?

Reply:
The problem is more about an error in the code which will be hard to diagnose afterwards because of the invalid pointer.